crates/kubelet/src/state/common/resources.rs

```diff
@@ -0,0 +1,127 @@
+//! Resources can be successfully allocated to the Pod.
+use crate::pod::state::prelude::*;
+use crate::provider::DevicePluginSupport;
+use crate::resources::device_plugin_manager::PodResourceRequests;
+use crate::resources::util;
+use crate::volume::{HostPathVolume, VolumeRef};
+use k8s_openapi::api::core::v1::HostPathVolumeSource;
+use k8s_openapi::api::core::v1::Volume as KubeVolume;
+use k8s_openapi::apimachinery::pkg::api::resource::Quantity;
+use std::collections::HashMap;
+use tracing::{debug, error, info};
+
+use super::error::Error;
+use super::image_pull::ImagePull;
+use super::{GenericPodState, GenericProvider};
+
+/// Resources can be successfully allocated to the Pod
+pub struct Resources<P: GenericProvider> {
+    phantom: std::marker::PhantomData<P>,
+}
+
+impl<P: GenericProvider> std::fmt::Debug for Resources<P> {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        "Resources".fmt(formatter)
+    }
+}
+
+impl<P: GenericProvider> Default for Resources<P> {
+    fn default() -> Self {
+        Self {
+            phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<P: GenericProvider> State<P::PodState> for Resources<P> {
+    async fn next(
+        self: Box<Self>,
+        provider_state: SharedState<P::ProviderState>,
+        pod_state: &mut P::PodState,
+        pod: Manifest<Pod>,
+    ) -> Transition<P::PodState> {
+        let pod = pod.latest();
+        debug!(pod = %pod.name(), "Preparing to allocate resources for this pod");
+        let device_plugin_manager = provider_state.read().await.device_plugin_manager();
+
+        // Only check for allocatable resources if a device plugin manager was provided.
+        if let Some(device_plugin_manager) = device_plugin_manager {
+            // Create a map of devices requested by this Pod's containers, keyed by container name
+            let mut container_devices: PodResourceRequests = HashMap::new();
+            for container in pod.all_containers() {
+                if let Some(resources) = container.resources() {
+                    if let Some(requests) = &resources.requests {
+                        let extended_resources: HashMap<String, Quantity> = requests
+                            .clone()
```

Comment:
Nit: Consider swapping this to `cloned` after the `filter` so you only clone the things you need

Reply:
I cant seem to use cloned without receiving this error: https://stackoverflow.com/questions/63132174/how-do-i-fix-the-method-clone-exists-but-the-following-trait-bounds-were-not

Reply:
Yep, then let's just skip that 😅